[RFC] vim-patch:7.4.{835,843,877}

jbradaric opened the pull request:
```
vim-patch:7.4.835

Problem:    Comparing utf-8 sequences does not handle different byte sizes
            correctly.
Solution:   Get the byte size of each character. (Dominique Pelle)
```

https://github.com/vim/vim/commit/f6470c288cb6f8efd60a507baf2c070f9d209ae6

```
vim-patch:7.4.843

Problem:    Still possible to go beyond the end of a string.
Solution:   Check for NUL also in second string. (Dominique Pelle)
```

https://github.com/vim/vim/commit/d43f0951bca162d4491d57df9277b5dbc462944f

```
vim-patch:7.4.877

Problem:    ":find" sometimes fails. (Excanoe)
Solution:   Compare current characters instead of previous ones.
```

https://github.com/vim/vim/commit/4d0c7bc74ac6fad5cb599dc3ade6996e848d83b6

jbradaric committed:
vim-patch:7.4.835

Problem:    Comparing utf-8 sequences does not handle different byte sizes
            correctly.
Solution:   Get the byte size of each character. (Dominique Pelle)

https://github.com/vim/vim/commit/f6470c288cb6f8efd60a507baf2c070f9d209ae6

jbradaric committed:
vim-patch:7.4.843

Problem:    Still possible to go beyond the end of a string.
Solution:   Check for NUL also in second string. (Dominique Pelle)

https://github.com/vim/vim/commit/d43f0951bca162d4491d57df9277b5dbc462944f

fwalch commented on src/nvim/file_search.c:
```diff
@@ -1067,18 +1067,19 @@ static int ff_wc_equal(char_u *s1, char_u *s2)
-    return FAIL;
-
-  for (i = 0; s1[i] != NUL && s2[i] != NUL; i += MB_PTR2LEN(s1 + i)) {
+  for (i = 0, j = 0; s1[i] != NUL;) {
```
You could pull the variable definitions for i and j into here.

fwalch commented on src/nvim/path.c:
```diff
@@ -1775,19 +1775,20 @@ bool same_directory(char_u *f1, char_u *f2)
   const char  *s = NULL;
 
-  for (i = 0; maxlen < 0 || i < maxlen; i += MB_PTR2LEN((char_u *)p + i)) {
+  for (i = 0, j = 0; maxlen < 0 || (i < maxlen && j < maxlen);) {
```
Same as above.

fwalch replied:
Same as above.

jbradaric replied:
These cannot be moved because `i` is used outside of the loop, when checking for a trailing slash.

fwalch commented:
Apart from minor comments, LGTM!

jbradaric committed:
file_search: Change return type of ff_wc_equal to bool.

This makes sense since the function returns only TRUE or FALSE.

jbradaric committed:
file_search: Declare loop variables inside the loop.

jbradaric commented:
@fwalch I will also add patch 7.4.843 to this PR because it seems that 7.4.835 didn't fix the problem completely.

jbradaric committed:
vim-patch:7.4.877

Problem:    ":find" sometimes fails. (Excanoe)
Solution:   Compare current characters instead of previous ones.

https://github.com/vim/vim/commit/4d0c7bc74ac6fad5cb599dc3ade6996e848d83b6

jbradaric commented:
I've also added patch 7.4.877 since it applies to the same part of the code as the previous patches.
